compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PropertyReferenceLowering.kt

```diff
@@ -258,120 +272,76 @@ internal class PropertyReferenceLowering(val context: JvmBackendContext) : Class
+                        }
                     }
-                    call.extensionReceiver = call.symbol.owner.extensionReceiverParameter?.let {
-                        if (expression.extensionReceiver != null)
-                            irGetField(irGet(dispatchReceiverParameter), receiverField)
-                        else
-                            irImplicitCast(irGet(valueParameters[index++]), it.type)
+                } else {
+                    fun IrBuilderWithScope.fieldReceiver(arguments: List<IrValueParameter>) = when {
+                        field.isStatic ->
+                            null
+                        expression.dispatchReceiver != null ->
+                            irImplicitCast(irGetField(irGet(arguments[0]), receiverField), field.parentAsClass.defaultType)
+                        else ->
+                            irImplicitCast(irGet(arguments[1]), field.parentAsClass.defaultType)
                     }
-                }
-
-                buildOverride(superClass.functions.single { it.name.asString() == "getName" }) { _, _ ->
-                    irString(expression.descriptor.name.asString())
-                }
-
-                buildOverride(superClass.functions.single { it.name.asString() == "getOwner" }) { _, _ ->
-                    buildReflectedContainerReference(expression)
-                }
-
-                buildOverride(superClass.functions.single { it.name.asString() == "getSignature" }) { _, _ ->
-                    irString(expression.signature)
-                }
 
-                expression.getter?.owner?.let { getter ->
-                    buildOverride(superClass.functions.single { it.name.asString() == "get" }) { dispatchReceiverParameter, valueParameters ->
-                        irGet(getter.returnType, null, getter.symbol).apply {
-                            copyTypeArgumentsFrom(expression)
-                            setReceiversOn(
-                                this,
-                                dispatchReceiverParameter,
-                                valueParameters
-                            )
-                        }
+                    referenceClass.addOverride(get!!) { arguments ->
+                        irGetField(fieldReceiver(arguments), field)
                     }
-                }
 
-                expression.setter?.owner?.let { setter ->
-                    buildOverride(superClass.functions.single { it.name.asString() == "set" }) { dispatchReceiverParameter, valueParameters ->
-                        val type = setter.valueParameters.last().type
-                        val value = irImplicitCast(irGet(valueParameters.last()), type)
-                        irSet(setter.returnType, null, setter.symbol, value).apply {
-                            copyTypeArgumentsFrom(expression)
-                            setReceiversOn(this, dispatchReceiverParameter, valueParameters)
+                    if (!field.isFinal) {
```

Comment:
Maybe best ```field.correspondingPropertySymbol.owner.isVar ```?

Reply:
correspondingPropertySymbol is null for Java fields.